### ISSUE TITLE
Campaign for the inactive transcoder

### ISSUE THREAD

**Issue by chrishobcroft** [livepeer/livepeer-monorepo#131]
**What is this discussion about? (required)**

I want to talk about how inactive transcoders are portrayed in the explorer.

**What do you hope to learn by the end of this discussion? (required)**

Some easy ways we can make inactive transcoders look less hopeless.

**Additional context**

![image](https://user-images.githubusercontent.com/2212651/41804475-6eb781ba-7697-11e8-9e75-7a58d117b45d.png)

In the example above, 2 of these nodes have been active and deactivated, 3 have never been active.

The 2 who once were active are displaying their reward cuts and fee shares.

The 3 who have never been active look like the most hopeless group, with zeroes everywhere... but some of them are willing to share a lot of their fees, and give away some of their reward.

[Heck, one of them is even offering to work for free _and_ give you all the money he makes](https://explorer.livepeer.org/accounts/0xccdc8000e5a9a79db949f8a4153971c72e0c76ca/transcoding)

Does anyone else think these guys are being represented unfavourably in the list of transcoders?

**Comment by ericxtang** [comment]
I think the path forward here is to use the ENS integration to add the campaigner's ENS identity.  @yondonfu worked on this [a few weeks ago](https://github.com/livepeer/livepeerjs/pull/113), still need to finish it up and merge it.

**Comment by dob** [comment]
1. I still stand by this issue: https://github.com/livepeer/livepeerjs/issues/33 that pending fee share/reward cut/price are more important to display than the active ones.

2. Reminder that even without an intense ENS integration that relies on full stack integration for not only player channel names but tons of transcoder metadata (that likely doesn't belong on chain), the data required to link transcoders to their campaigns (and even display names) is already available for the dapp to pull from here: https://github.com/dob/livepeer-lists/tree/master/transcoders

**Comment by chrishobcroft** [comment]
Regarding point 2, is this list only the active transcoders?

**Comment by dob** [comment]
It’s anyone who’s submitted a campaign or a PR to add a file for a new transcoder. I try to keep up and will add the new ones but anyone can submit a PR to add or update one.


________________________________

**Comment by jozanza** [comment]
@dob on point 1, #33 is actually the solution for this issue as well. I'll leave this one open until that one gets closed 👍 

Point 2 is entirely different from this issue. And as I've mentioned before, using [fixtures](https://github.com/dob/livepeer-lists/tree/master/transcoders) is not a sound way to build out  feature or enhancement related to transcoder identity or meta-data. This is not a static site. To store data, users must authenticate and that data must be read at runtime from remote storage. Importing a bunch of JSON files (that don't even live in the same repo) bloats the bundle, reduces maintainability, and introduces security issues. I'm not for it. 

ENS for names are great, and that's one piece of the this puzzle (There's already a [media player PR for this](https://github.com/livepeer/livepeerjs/pull/113)). But how to store and retrieve other meta data needs some real thought and a real solution. With a little thought and effort, we can do this properly.

**Comment by dob** [comment]
@jozanza if you treat the JSON files available in github as an API (called via HTTP requests), how is reading the data from these github hosted files different than hitting the etherscan API or the infura api to read data from other sources?

**Comment by jozanza** [comment]
It's using the same transport as etherscan or infura (HTTP), and that's about all. If you use raw.github instead of fixtures, that's slightly better, but still not great. Here's why:

#### 1. Not Open; Poor UX
  - Private feature for Github users, but the dApps are not developer-specific. Storing meta-data is useful feature for all users
  - Might as well "Login with Github" and use the API if you go that route, but not all users can/will/should login with Github.
  - Are we going to make a new repo each time we need new meta info? How do we let folks know about schema updates are new fields? Do we link them to create Github PRs?
  - We also need meta info beyond the context of just the Explorer: it's needed in multiple dApps and contexts (Media Player, CLI, APIs, SDK, etc). This is not a one-off feature -- it's core infrastructure for the JavaScript stack.

#### 2. Security Problem
  - Can't easily validate PR authors as ETH account private key owners
  - Manual approach to validation is time-consuming and error-prone
  - Automated approach -- should we really take the time to setup CI hooks and build dApps around this pattern?

#### 3. Centralized
  - We own it; We're the gatekeepers; Freshness of data also depends on our bandwidth for reviewing/accepting PRs.
  - If the purity of our dApp becomes an issue, we can easily replace Etherscan/Infura with geth node or manual tx object resolution from the chain itself. But moving away from Github raw files is not so simple. And it will affect the user as they will likely have to go through extra steps, such as signing or submitting a transaction.
  - App developers using the SDK would also have to piggyback off of raw.github for this info as well.

#### 4. Removes one of the most interesting parts of the modern P2P stack
  - JS has great libs for working with decentralized storage (dat, libp2p, etc). Devs will be far more interested in a project actually using that tech.
  - The Explorer and Media Player serve as examples of well-built production-quality dApps. Using the Github raw API is probably more suitable for weekend projects and thus lowers that standard.
  - The PR approach won't scale well at all; As the transcoder pool and frequency of updates grows, it will become unmanageable.

**tl;dr** We need a way to do this in-app / programmatically.

**Comment by dob** [comment]
I of course recognize my suggestion is not the ideal long term engineering solution for how we associate metadata in a decentralized way. I just want to clarify why I'm even suggesting this in the first place:

1. I'm thinking about the user - it's been 60 days since the protocol launched, and there's been no link between the transcoder's appearance on the election dashboard and their social campaign. This would be really helpful to a user however it's done!

2. I believe very much in the classic Paul Graham advice of [Do Things That Don't Scale](http://paulgraham.com/ds.html) (see the Manual section), in the early days of a startup before fit and rapid growth. Hustle to deliver user value based upon what you hear from them.

3. If you look at what's actually going on here, there's been exactly 13 campaign links in the 60 days since network launched. None of them have ever been updated (as in the URL has changed after being posted). So the way to deliver tremendous missing value to our delegators is to add 13 links to a page. This seems like a perfect opportunity for "doing things that don't scale" for the time being to me. 

If super concerned about the architecture (as you rightfully should be!) then there are steps we can take to make this replaceable with a proper decentralized store in the future, such as hiding the implementation behind an instance of a MetaDataProvider API. The raw.github implementation could work, but we could even do a hardcoded implementation, or a read from a single IPFS hash or s3 location implementation.

Again, just sharing my motivation and thought process here. I of course respect standing up for proper engineering practices, so I would not suggest sacrificing the integrity of the app. I'm all about giving this value to a user asap rather than making them wait another 60 days before connecting transcoder campaigns to links.

**Comment by chrishobcroft** [comment]
It's always a fun one, this "purist vs. pragmatist" trade-off.

I see both sides, and would like to know how doing the things that don't scale would degrade a user's experience of interacting with Livepeer? Perhaps that's one for @jozanza?

And one potentially pertinent point here is to make a correction to point 3 above. The URLs of Transcoder Campaigns *have* changed since being first published - they change when the title of the post is updated.

![screenshot_20180629-011959](https://user-images.githubusercontent.com/2212651/42065576-02a7a5fa-7b3c-11e8-8bdc-a881cc4dc792.png)

Also, given that anyone (I think, or maybe only an admin) can edit this, it could end up as a nightmare in maintenance.

I guess this is a little niggly example of something where maintaining the same data in disparate systems becomes an overhead, with potential for a 404, although I'm even inclined to agree that connecting transcoders with their campaigns (and even with their broadcast channel).

**Comment by chrishobcroft** [comment]
Thinking a little outside the box here though - surely the right answer is to embed the media player in the broadcaster tab for a node.

This would allow Transcoder Nodes to stream content relating to their node's campaign, which is much more "blockchainy" than a webpage with lots of text and images, IMHO.

Also would be a good way to attract more inbound traffic to the explorer in general.

**Comment by chrishobcroft** [comment]
Interested to know how this one has been solved.

**Comment by jozanza** [comment]
This talk started out as a discussion about displaying inactive transcoder metrics fairly. That is solved (along with #33) by [this PR](https://github.com/livepeer/livepeerjs/commit/bb0549f698911c471f70223fd77b7745f4be698c).

Later on this thread got off-topic and focused on the challenge of displaying transcoder meta information, which is already an issue: https://github.com/livepeer/livepeerjs/issues/100

Feel free to continue that convo there.
